website/blog/2019-12-19-improve-presto-planner.md

```diff
@@ -0,0 +1,129 @@
+
+`PickTableLayout` and `AddExchanges` are two of the more important optimizers.
+
+`PickTableLayout` plans predicate push down into a table scan by calling the connector provided API method `getTableLayout`. It is also used to obtain physical layout information from the connector. `getTableLayout` returns a `LayoutHandle` that the connector populates with information on the structure of the data that will be returned by the scan. Presto will later use the `LayoutHandle` to plan, optimize, and execute the query.
+
+`AddExchanges` adds the data shuffling (data exchange) operators to a query execution. This important step determines how query execution is parallelized, and how data is redistributed for processing at each stage of a query. A stage of execution in Presto is generally the shuffling of data on the partition key that is required for processing the next part of a query plan. `AddExchanges` relies on the Handle returned from the connector to decide on the appropriate places and kinds of exchanges to add to the plan.
+
+Relying on `PickTableLayout` to do both predicate push down and physical planning is very restrictive because there is no way for connectors to modify the plan beyond basic predicate pushdown.
+
+Presto now allows connectors to provide optimization rules to the Presto engine, which allows Connectors to introduce arbitrary optimizations. There are restrictions to prevent Connector provided optimizers from accidentally changing another Connector’s sub plan:
+1.  PlanNodes that are exposed to presto-spi module.
+2.  PlanNodes that belong to the connector.
+
+A sub max tree that satisfies the above rules will be transformed to more optimized form picked by a connector provided optimization rule:
+```java
+public  interface  ConnectorPlanOptimizer
+{
+	PlanNode optimize(
+		PlanNode maxSubplan,
+		ConnectorSession session,
+		VariableAllocator variableAllocator,
+		PlanNodeIdAllocator idAllocator);
+}
+```
+![Remote Exchange](/img/blog/2019-12-19-improve-presto-planner.md/image_03.png)
+
+The connector rule will transform the sub max plan tree. In the case of predicate pushdown (taking `MySQLConnector` as an example), the connector can save the predicates that MySQL can handle as a SQL expression inside `MySQLConnectorLayoutHandle` and return a `TableScan` node.
+
+The connector provided rule can not only remove/modify plan nodes (push down) that belong to the connector, but also add new nodes belonging to another connector (view expansion). In the latter case, TableScan might be reading from a virtual table combining data from multiple different data sources. The TableScan on the virtual table can be expanded to a new sub tree that unions TableScans from both sides. Once expanded, the optimization of newly generated plan nodes can then be handled by the connectors they belong to, so that most optimized subplan can be achieved.
```

Comment:
This contradicts what you just said where a connector can only modify its own plan nodes. Is there a way to reconcile the two?

Reply:
I think this paragraph means a connector A can "spawn" another one B. But once B spawned, A will lose the control of B and leaving B's optimizer to take the control.

Reply:
@aweisberg , reorganized a little bit, hopes it makes it clearer for the reader now.